### DISCUSSION TITLE
WinUI alternative for Wpf's CreateBitmapSourceFromHIcon.

### DISCUSSION THREAD

**Discussion by gautambjain** [microsoft/microsoft-ui-xaml#7943] [Q&A] [answered]
Wpf had Imaging.CreateBitmapSourceFromHIcon . How do I do it in WinUI?

Basically, I want to show a file's icon in a listview. I plan to use SHGetFileInfo to get HICON. But don't know how to show it in list view.

Please help :o)

**Comment by castorix** [comment, marked as answer]
> Basically, I want to show a file's icon in a listview.I plan to use SHGetFileInfo to get HICON. 

I think it is simpler to use [GetThumbnailAsync](https://learn.microsoft.com/en-us/uwp/api/windows.storage.storagefile.getthumbnailasync?view=winrt-22621), then you can set it as BitmapSource for a control
(like in the test I did in this thread : [Set System.Drawing.Icon as Source for ImageIcon control - WinUI 3](https://learn.microsoft.com/en-us/answers/questions/1037472/set-systemdrawingicon-as-source-for-imageicon-cont.html))

**Reply by gautambjain** [reply]
Thanks again @castorix for rapid answers. God bless you.

**Reply by gautambjain** [reply]
This is working. The above method does 2 things - 

1) For regular non-image files, it loads the file's icon (of associated app) like shown in Windows File Explorer 
2) For image files (jpg, png etc), it loads the thumbnail (small version of the image itself). Which affects the performance of my app.

I don't want the thumbnails to be loaded. I just want the file's icon so that the files list in my app loads faster.

Is there a way for this?

**Reply by castorix** [reply]
> I don't want the thumbnails to be loaded. I just want the file's icon so that the files list in my app loads faster.
> Is there a way for this?

Once you got the icon handle, you can convert it to _WriteableBitmap_
Instead of _SHGetFileInfo_, MS recommends [IExtractIcon  ](https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nn-shlobj_core-iextracticonw), but it needs more code
To get a WriteableBitmap, it can be done with GDI, like I did in the _GetCaptureWriteableBitmap_ function (https://github.com/castorix/WinUI3_MediaEngine/blob/master/CMediaEngine.cs)
You can use GetIconInfo to convert first hIcon to hBitmap (I tested with Notepad icon and returned _hbmColor_, then the returned WriteableBitmap set to .Source of an Image control)

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool GetIconInfo(IntPtr hIcon, out ICONINFO piconinfo);

        [StructLayout(LayoutKind.Sequential)]
        public struct ICONINFO
        {
            public bool fIcon;
            public uint xHotspot;
            public uint yHotspot;
            public IntPtr hbmMask;
            public IntPtr hbmColor;
        }

**Reply by gautambjain** [reply]
Fantastic! This worked.

**Comment by Slion** [comment]
You could also do something like that:



```C#
using System.Drawing;
using System.Drawing.Imaging;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Xaml.Media.Imaging;

  // Use GDI+ to load our HICON
  Bitmap bmp = Icon.FromHandle(hIcon).ToBitmap();
  using var ms = new MemoryStream();
  // Save it as PNG into RAM
  bmp.Save(ms, ImageFormat.Png);
  ms.Seek(0, SeekOrigin.Begin);
  // Create and loud our bitmap source
  var bitmapImage = new BitmapImage();
  // TODO: Use async variant
  bitmapImage.SetSource(ms.AsRandomAccessStream());

  // Use GDI+ to load our HBITMAP
  Bitmap bmp = Image.FromHbitmap(hBitmap);
  using var ms = new MemoryStream();
  // Save it as PNG into RAM
  bmp.Save(ms, ImageFormat.Png);
  ms.Seek(0, SeekOrigin.Begin);
  // Create and load our bitmap source
  var bitmapImage = new BitmapImage();
  // TODO: Use async variant
  bitmapImage.SetSource(ms.AsRandomAccessStream());
```
You will need to use NuGet System.Drawing.Common.
I'm still using NuGet System.Drawing.Common v8.0.8 as I recall the 9.x did not work with WinUI for some reason.
I came up with it from looking at the following:
https://stackoverflow.com/a/41999179/3969362
https://stackoverflow.com/a/33221178/3969362
